OpenRA.Mods.Common/Traits/Buildings/DockHost.cs

```diff
@@ -0,0 +1,174 @@
+{
+	public sealed class DockType { DockType() { } }
+
+	[Desc("A generic dock that services DockClients.")]
+	public class DockHostInfo : ConditionalTraitInfo, Requires<RefineryInfo>
+	{
+		[Desc("Docking type.")]
+		public readonly BitSet<DockType> Type;
+
+		[Desc("How many clients can this dock be reserved for?")]
+		public readonly int MaxQueueLength = 3;
+
+		[Desc("How long should the harvester wait before starting the docking sequence.")]
+		public readonly int DockWait = 10;
+
+		[Desc("Actual client facing when docking.")]
+		public readonly WAngle DockAngle = WAngle.Zero;
+
+		[Desc("Docking cell relative to the centre of the actor.")]
+		public readonly WVec DockOffset = WVec.Zero;
+
+		[Desc("Does client need to be dragged in?")]
+		public readonly bool IsDragRequired = false;
+
+		[Desc("Vector by which the client will be dragged when docking.")]
+		public readonly WVec DragOffset = WVec.Zero;
+
+		[Desc("In how many steps to perform the dragging?")]
+		public readonly int DragLength = 0;
+
+		public override object Create(ActorInitializer init) { return new DockHost(init.Self, this); }
+	}
+
+	public class DockHost : ConditionalTrait<DockHostInfo>, ITick, INotifySold, INotifyCapture, INotifyOwnerChanged, ISync, INotifyKilled, INotifyActorDisposing
+	{
+		readonly Actor self;
+
+		public BitSet<DockType> GetDockType => Info.Type;
+		public bool IsEnabledAndInWorld => !preventDock && !IsTraitDisabled && !self.IsDead && self.IsInWorld;
+		public int ReservationCount => reservedDockClients.Count;
+		public bool CanReserve => ReservationCount < Info.MaxQueueLength;
+		readonly List<DockClientManager> reservedDockClients = new List<DockClientManager>();
+
+		public WPos DockPosition => self.CenterPosition + Info.DockOffset;
+		public int DockWait => Info.DockWait;
+		public WAngle DockAngle => Info.DockAngle;
+		public bool IsDragRequired => Info.IsDragRequired;
+		public WVec DragOffset => Info.DragOffset;
+		public int DragLength => Info.DragLength;
+
+		[Sync]
+		bool preventDock = false;
+
+		[Sync]
+		Actor dockedClientActor = null;
+		DockClientManager dockedClient = null;
+		readonly Func<Actor, DockClientManager, Actor, DockHost, Activity> dockSequence;
+
+		public DockHost(Actor self, DockHostInfo info)
+			: base(info)
+		{
+			this.self = self;
+
+			// HACK: Remove once TiberianSunRefinery is disposed of.
+			dockSequence = self.Trait<Refinery>().DockSequence;
+		}
+
+		/// <summary> Can this client dock at this host. </summary>
+		/// <remarks>
+		/// Does not check <see cref="DockType"/>.
+		/// Does not check if <see cref="IDockClient"/> is enabled.
+		/// Does not check if <see cref="DockClientManager"/> is enabled.
+		/// </remarks>
+		public virtual bool DockingPossible(Actor clientActor, IDockClient client, bool ignoreReservations = false)
+		{
+			return !IsTraitDisabled && clientActor.Owner.IsAlliedWith(self.Owner) && (ignoreReservations || CanReserve || reservedDockClients.Contains(client.DockClientManager));
```

Comment:
We should check if Allied docking can be used, as this is something that they should be able to disable easily.

```C#
clientActor.Owner.IsAlliedWith(self.Owner)
```
 becomes
```C#
(clientActor.Owner.IsAlliedWith(self.Owner) || AlliedDockingDisabled)
```

Reply:
It would have to be
```c#
AlliedDockingDisabled) ? self.Owner == clientActor.Owner : (clientActor.Owner.IsAlliedWith(self.Owner)
```

I suppose I could move that check elsewhere. It doesn't feel like it belongs with these other fairly abstract checks

Reply:
Aren't players allied with themselves by default? Though you're right, checking strictly for equivalence is cleaner.